### PR TITLE
Split ZoneOffset into UtcOffset and FixedOffsetTimeZone

### PR DESCRIPTION
**Pull request by ilya-g**
This PR separates the concept of a time zone with a fixed UTC offset and from the concept of UTC offset itself.

Remaining task items:
- [x] serialization of UtcOffset
- [x] decide on UtcOffset normalization
- [ ] ~make UtcOffset comparable~
  - [ ] ~decide on the default ordering~

Work postponed for a further PR:
- documentation of new public API
- Allow non-normalized fixed offset time zone, such as `UTC+04` or `Etc/UTC`, to be represented by `FixedOffsetTimeZone` preserving its `id`.

### CODE REVIEW

File: core/build.gradle.kts [hunk 1]
```diff
@@ -22,7 +22,7 @@ val JDK_8: String by project
 val serializationVersion: String by project
 
 kotlin {
-    explicitApi()
+    explicitApiWarning()
```
Comment (ilya-g) [inline core/build.gradle.kts:25]: todo: rollback this change

File: core/common/src/TimeZone.kt [hunk 2]
```diff
@@ -85,18 +86,36 @@ public expect open class TimeZone {
...
+
+@Deprecated("Use FixedOffsetTimeZone of UtcOffset instead", ReplaceWith("FixedOffsetTimeZone"))
+public typealias ZoneOffset = FixedOffsetTimeZone
+
+@Serializable(with = UtcOffsetSerializer::class)
+public expect class UtcOffset {
     public val totalSeconds: Int
+
+    public companion object {
+        public fun parse(offsetString: String): UtcOffset
```
Comment (ilya-g) [inline core/common/src/TimeZone.kt:106]: Parameter name, how would it be better to call it?
- offsetString
- offsetId
- offset
- isoString
- just text as in some other `parse` functions

File: core/common/src/serializers/TimeZoneSerializers.kt [hunk 3]
```diff
@@ -23,21 +24,35 @@ public object TimeZoneSerializer: KSerializer<TimeZone> {
...
     }
 
-    override fun serialize(encoder: Encoder, value: ZoneOffset) {
+    override fun serialize(encoder: Encoder, value: FixedOffsetTimeZone) {
         encoder.encodeString(value.id)
     }
 
 }
+
+public object UtcOffsetSerializer: KSerializer<UtcOffset> {
```
Comment (ilya-g) [inline core/common/src/serializers/TimeZoneSerializers.kt:46]: We could also provide an alternative serializer that stores offset as an integer primitive or a single component object.

File: core/jvm/src/TimeZoneJvm.kt [hunk 4]
```diff
@@ -51,24 +52,52 @@ public actual open class TimeZone internal constructor(internal val zoneId: Zone
...
+} catch (e: DateTimeException) {
+    throw DateTimeArithmeticException(e)
+}
+
+
 public actual fun LocalDateTime.toInstant(timeZone: TimeZone): Instant =
         this.value.atZone(timeZone.zoneId).toInstant().let(::Instant)
 
+public actual fun LocalDateTime.toInstant(utcOffset: UtcOffset): Instant =
+        this.value.atZone(utcOffset.zoneOffset).toInstant().let(::Instant)
```
Comment (ilya-g) [inline core/jvm/src/TimeZoneJvm.kt:100]: Can be simplified to `this.value.toInstant(utcOffset.zoneOffset).let(::Instant)`

File: core/native/src/Instant.kt [hunk 5]
```diff
@@ -234,22 +234,31 @@ public actual class Instant internal constructor(public actual val epochSeconds:
...
 
-private fun Instant.toZonedLocalDateTimeFailing(zone: TimeZone): ZonedDateTime = try {
-    toZonedLocalDateTime(zone)
+private fun Instant.toZonedDateTimeFailing(zone: TimeZone): ZonedDateTime = try {
+    toZonedDateTime(zone)
 } catch (e: IllegalArgumentException) {
     throw DateTimeArithmeticException("Can not convert instant $this to LocalDateTime to perform computations", e)
 }
 
+// org.threeten.bp.LocalDateTime#ofEpochSecond + org.threeten.bp.ZonedDateTime#create
```
Comment (dkhalanskyjb) [inline core/native/src/Instant.kt:243]: This comment now makes little sense (`LocalDateTime#ofEpochSeconds` is now used in `Instant.toLocalDateTimeImpl`, not here). Also, maybe inline this function?

File: core/native/src/TimeZone.kt [hunk 6]
```diff
@@ -80,137 +91,46 @@ public actual open class TimeZone internal constructor(internal val value: TimeZ
...
-        }
 
-        // org.threeten.bp.ZoneOffset#ofHoursMinutesSeconds
-        internal fun ofHoursMinutesSeconds(hours: Int, minutes: Int, seconds: Int): ZoneOffset {
-            validate(hours, minutes, seconds)
-            return if (hours == 0 && minutes == 0 && seconds == 0) UTC
-            else ofSeconds(hours * SECONDS_PER_HOUR + minutes * SECONDS_PER_MINUTE + seconds)
-        }
+public actual fun TimeZone.offsetAt(instant: Instant): UtcOffset =
+        value.offsetAt(instant)
```
Comment (dkhalanskyjb) [inline core/native/src/TimeZone.kt:109]: Minor: inconsistent indentation.

**Review by dkhalanskyjb — CHANGES_REQUESTED** [review]

File: core/jvm/src/Converters.kt [hunk 7]
```diff
@@ -61,11 +61,22 @@ public fun java.time.ZoneId.toKotlinTimeZone(): TimeZone = TimeZone(this)
...
+ * Converts this [kotlinx.datetime.FixedOffsetTimeZone][FixedOffsetTimeZone] value to a [java.time.ZoneOffset][java.time.ZoneOffset] value.
  */
-public fun ZoneOffset.toJavaZoneOffset(): java.time.ZoneOffset = this.zoneOffset
+public fun FixedOffsetTimeZone.toJavaZoneOffset(): java.time.ZoneOffset = this.offset.zoneOffset
 
 /**
- * Converts this [java.time.ZoneOffset][java.time.ZoneOffset] value to a [kotlinx.datetime.ZoneOffset][ZoneOffset] value.
+ * Converts this [java.time.ZoneOffset][java.time.ZoneOffset] value to a [kotlinx.datetime.FixedOffsetTimeZone][FixedOffsetTimeZone] value.
  */
-public fun java.time.ZoneOffset.toKotlinZoneOffset(): ZoneOffset = ZoneOffset(this)
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/jvm/src/Converters.kt:71]: Yes, makes sense

File: core/js/src/UtcOffset.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import kotlinx.datetime.serializers.UtcOffsetSerializer
+import kotlinx.serialization.Serializable
+
+@Serializable(with = UtcOffsetSerializer::class)
+public actual class UtcOffset(internal val zoneOffset: ZoneOffset) {
+    public actual val totalSeconds: Int get() = zoneOffset.totalSeconds().toInt()
+
+    override fun hashCode(): Int = zoneOffset.hashCode().toInt()
+    override fun equals(other: Any?): Boolean = other is UtcOffset && this.zoneOffset == other.zoneOffset
+    override fun toString(): String = zoneOffset.toString()
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/js/src/UtcOffset.kt:18]: I'd keep ZoneOffset instance for interop with js-joda library.

File: core/common/test/TimeZoneTest.kt [hunk 9]
```diff
@@ -93,73 +103,96 @@ class TimeZoneTest {
...
 
+    @Test
+    fun utcOffsetNormalization() {
+        val sameOffsetTZs = listOf("+04", "+04:00", "UTC+4", "UT+04").map { TimeZone.of(it) }
+        val instant = Instant.fromEpochSeconds(0)
+        val offsets = sameOffsetTZs.map { it.offsetAt(instant) }
+
+        assertTrue(offsets.distinct().size == 1, "Expected all offsets to be equal: $offsets")
+        assertTrue(offsets.map { it.toString() }.distinct().size == 1, "Expected all offsets to have the same string representation: $offsets")
+    }
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/common/test/TimeZoneTest.kt:114]: Offsets of differently named time zones with the same fixed offset are equal.

File: core/common/test/TimeZoneTest.kt [hunk 10]
```diff
@@ -55,7 +59,13 @@ class TimeZoneTest {
 
         assertFailsWith<IllegalTimeZoneException> { TimeZone.of("Mars/Standard") }
         assertFailsWith<IllegalTimeZoneException> { TimeZone.of("UTC+X") }
+    }
 
+    @Test
+    fun ofFailsOnInvalidOffset() {
+        for (v in UtcOffsetTest.invalidUtcOffsetStrings) {
+            assertFailsWith<IllegalTimeZoneException> { TimeZone.of(v) }
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/common/test/TimeZoneTest.kt:67]: Yes, but not in this PR

File: core/common/test/InstantTest.kt [hunk 11]
```diff
@@ -630,3 +630,13 @@ inline fun <T> assertInvalidFormat(message: String? = null, f: () -> T) {
     }
 }
 
+
+
+@Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER")
+@kotlin.internal.InlineOnly
+inline fun <T> assertIllegalArgument(message: String? = null, f: () -> T) {
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/common/test/InstantTest.kt:637]: All of them should be moved then.

File: core/js/src/TimeZone.kt [hunk 12]
```diff
@@ -28,12 +27,12 @@ public actual open class TimeZone internal constructor(internal val zoneId: Zone
 
     public actual companion object {
         public actual fun currentSystemDefault(): TimeZone = ZoneId.systemDefault().let(::TimeZone)
-        public actual val UTC: TimeZone = jtZoneOffset.UTC.let(::TimeZone)
+        public actual val UTC: FixedOffsetTimeZone = UtcOffset(jtZoneOffset.UTC).asTimeZone()
 
         public actual fun of(zoneId: String): TimeZone = try {
             val zone = ZoneId.of(zoneId)
             if (zone is jtZoneOffset) {
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/js/src/TimeZone.kt:34]: Any examples of such time zones?

File: core/js/src/UtcOffset.kt [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+public actual fun UtcOffset(hours: Int = 0, minutes: Int = 0, seconds: Int = 0): UtcOffset =
+    try {
+        if (hours == 0 && minutes == 0) {
+            UtcOffset(ZoneOffset.ofTotalSeconds(seconds))
+        } else {
+            UtcOffset(ZoneOffset.ofHoursMinutesSeconds(hours, minutes, seconds))
+        }
+    } catch (e: Throwable) {
+        if (e.isJodaDateTimeException()) throw IllegalArgumentException(e) else throw e
+    }
```
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/js/src/UtcOffset.kt:43]: It's interesting to note that java.time doesn't have a function like `ofTotalMinutes()`, yet there are no questions about how to construct offset from minutes. Maybe it's an evidence that it's not actually needed, or that it's obvious that `ofTotalSeconds(minutes * 60)` can be used instead.

File: core/common/test/TimeZoneTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/common/test/TimeZoneTest.kt:114]: Still don't see why this is useful. If these time zones are not considered equal, then they are just a group of separate time zones. It's like checking that today Europe/Moscow and Europe/Helsinki happen to have the same offset. True, but so what?

I guess this test checks that the equality between offsets does not depend on the timezone in which the offset was queried?

File: core/js/src/TimeZone.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/TimeZone.kt:34]: Not that I know of.

File: core/js/src/UtcOffset.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/UtcOffset.kt:43]: Sure, but the difference is that Java's approach is consistent: they provide a group of functions that behave in an expected way, though, yes, some functions that could have been implemented are missing. Discontinuity, in my opinion, is bad regardless of whether it can be worked around and whether it is triggered commonly. I could elaborate with specific examples of potential problems this discontinuity could cause if you'd like me to.

File: core/js/src/UtcOffset.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/UtcOffset.kt:18]: js-joda caches `ZoneOffset` instances: https://github.com/js-joda/js-joda/blob/master/packages/core/src/ZoneOffset.js#L287, so I don't think constructing `ZoneOffset` on the fly in the converter functions could become a bottleneck.

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/UtcOffset.kt:18]: Same goes for the Java implementation.

File: core/common/test/TimeZoneTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/common/test/TimeZoneTest.kt:114]: This test is specifically for fixed-offset time zones. After the further improvement in another branch, the offset will be queried without an instant, e.g. https://github.com/Kotlin/kotlinx-datetime/commit/0cc11b025c2e308670b7d6936ea5e55bdf9c985a

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/common/test/TimeZoneTest.kt:114]: Ok, I found out what non-trivial behavior is checked by this test but not by any other: the fact that time zone identifiers of these forms are parsed correctly. Maybe also add `GMT` then?

File: core/js/src/UtcOffset.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/js/src/UtcOffset.kt:43]: I feel that this constructor function contract is already overloaded with two ways of checking parameter ranges, and not sure that it's worth adding another way to the mix.

Please share your concerns about discontinuity.

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/UtcOffset.kt:43]: This contract could be simplified to just "everything that comes after the first non-zero component gets range-checked". Seems succinct. Or we could prohibit the `Utc(seconds = 7200)` usage for now and possibly add it in some form if there's demand.

As for concerns, consider this hypothetical. Let's say a user attempts to use this function with 180 minutes. The system complains, talking about range-checks. The user infers that, surely, this means that this function behaves in the same way as `ofHoursMinutesSeconds` in Java, that is, it expects the arguments to reflect the human-readable representation of the offset. Then, encountering the form `UtcOffset(seconds = 7200)`, they believe that it's a bug. And much, much smaller things have led to users losing their sense of flow and becoming greatly confused.

tl;dr: other APIs for `UtcOffset` equivalents and for similar things (like, in a sense, `LocalTime`) provide constructors that range-check, and the user is at risk of classifying this constructor as such if they happen to encounter range-checking of minutes.

[the remark this answers is not in the document]
Comment (ilya-g) [reply core/js/src/UtcOffset.kt:43]: I agree that the proposed change would simplify the contract of this constructor, but this comes at the expense of not enforcing the same value ranges as `ofHoursMinutesSeconds` in Java or the `UtcOffset.parse` function do.
So for example, if a user has parsed the components of an offset like `00:60:00` manually and now wants to construct an offset validating the range of these components, `UtcOffset` function would be of no help here.

The problem lies in the fact that the missing hours component is indistinguishable from the present hours that is equal to zero. This brings an idea that we could use nullable Int parameters and `null` as a default value to distinguish these two situations.

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/UtcOffset.kt:43]: > So for example, if a user has parsed the components of an offset like `00:60:00` manually and now wants to construct an offset validating the range of these components, `UtcOffset` function would be of no help here.

The same can be said about `00:00:60`. That's the problem with the discontinuity: minutes and seconds are not inherently different, and if we behave like they are, it's inconsistent and confusing.

> This brings an idea that we could use nullable Int parameters and `null` as a default value to distinguish these two situations.

Could you clarify the behavior that you invision here? I don't understand what `null` would mean. Also, which parameters would be nullable?

[the remark this answers is not in the document]
Comment (ilya-g) [reply core/js/src/UtcOffset.kt:47]: Basically it means the same thing as you proposed "everything that comes after the first non-zero component gets range-checked", except of "non-null" instead of "non-zero".

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply core/js/src/UtcOffset.kt:47]: This looks semantically clean. I don't know if such a solution is idiomatic in Kotlin, but you're a better judge of that than I am.

File: core/common/test/TimeZoneTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (ilya-g) [reply core/common/test/TimeZoneTest.kt:114]: Ok, added one

File: core/common/src/TimeZone.kt (hunk 2 above)
Comment (ilya-g) [reply core/common/src/TimeZone.kt]: Postpone decision, review all such functions together.

File: core/common/src/UtcOffset.kt [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+@Serializable(with = UtcOffsetSerializer::class)
+public expect class UtcOffset {
+    public val totalSeconds: Int
+
+    public companion object {
+        public val ZERO: UtcOffset
+        public fun parse(offsetString: String): UtcOffset
+    }
+}
+public expect fun UtcOffset(hours: Int? = null, minutes: Int? = null, seconds: Int? = null): UtcOffset
```
Comment (dkhalanskyjb) [inline core/common/src/UtcOffset.kt:20]: This needs to be documented. The other new functions as well, but the rest of them are much more self-explanatory.
